tls/extensions/s2n_client_key_share.c

```diff
@@ -191,7 +219,47 @@ static int s2n_send_hrr_pq_hybrid_keyshare(struct s2n_connection *conn, struct s
     notnull_check(conn);
     notnull_check(out);
 
-    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+    /* If in FIPS mode, the client should not have sent any PQ IDs
+     * in the supported_groups list of the initial ClientHello */
+    ENSURE_POSIX(s2n_is_in_fips_mode() == false, S2N_ERR_PQ_KEMS_DISALLOWED_IN_FIPS);
+
+    const struct s2n_ecc_preferences *ecc_pref = NULL;
+    GUARD(s2n_connection_get_ecc_preferences(conn, &ecc_pref));
+    notnull_check(ecc_pref);
+
+    const struct s2n_kem_preferences *kem_pref = NULL;
+    GUARD(s2n_connection_get_kem_preferences(conn, &kem_pref));
+    notnull_check(kem_pref);
+
+    const struct s2n_kem_group *server_negotiated_kem_group = conn->secure.server_kem_group_params.kem_group;
+    ENSURE_POSIX(server_negotiated_kem_group != NULL, S2N_ERR_INVALID_HELLO_RETRY);
+    ENSURE_POSIX(s2n_kem_preferences_includes_tls13_kem_group(kem_pref, server_negotiated_kem_group->iana_id),
+            S2N_ERR_INVALID_HELLO_RETRY);
+
+    for (size_t i = 0; i < kem_pref->tls13_kem_group_count; i++) {
+        if (kem_pref->tls13_kem_groups[i]->iana_id == server_negotiated_kem_group->iana_id) {
+            struct s2n_kem_group_params *kem_group_params = &conn->secure.client_kem_group_params[i];
+            ENSURE_POSIX(kem_group_params->kem_group == NULL, S2N_ERR_INVALID_HELLO_RETRY);
+            ENSURE_POSIX(kem_group_params->ecc_params.evp_pkey == NULL, S2N_ERR_INVALID_HELLO_RETRY);
+            ENSURE_POSIX(kem_group_params->kem_params.private_key.data == NULL, S2N_ERR_INVALID_HELLO_RETRY);
+            kem_group_params->kem_group = kem_pref->tls13_kem_groups[i];
+            GUARD(s2n_generate_pq_hybrid_key_share(conn, out, kem_group_params));
+        } else if (conn->secure.client_kem_group_params[i].kem_group != NULL) {
+            /* Wipe any key shares that weren't chosen for negotiation */
+            GUARD(s2n_kem_group_free(&conn->secure.client_kem_group_params[i]));
+            conn->secure.client_kem_group_params[i].kem_group = NULL;
+            conn->secure.client_kem_group_params[i].kem_params.kem = NULL;
+            conn->secure.client_kem_group_params[i].ecc_params.negotiated_curve = NULL;
+        }
+    }
+
+    /* Wipe any ECC shares that may have been sent, since PQ was selected for negotiation */
+    for (size_t i = 0; i < ecc_pref->count; i++) {
+        GUARD(s2n_ecc_evp_params_free(&conn->secure.client_ecc_evp_params[i]));
+        conn->secure.client_ecc_evp_params[i].negotiated_curve = NULL;
+    }
```

Comment:
Would it be cleaner to just wipe all key shares (after checking the new key share is empty) before calculating the new key share, so you can use the same wiping logic for both PQ and non-PQ?